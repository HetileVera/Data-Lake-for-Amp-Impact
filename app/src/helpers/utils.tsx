import isEmpty from "lodash/isEmpty";
import isNil from "lodash/isNil";
import isString from "lodash/isString";
import numeral from "numeral";

export const delay = (seconds: number) => {
  return new Promise((resolve) => {
    setTimeout(resolve, seconds * 1000);
  });
};

export function humanDay(num: number): string {
  // Account for x11 to x13
  const remainder100 = num % 100;
  if (remainder100 >= 11 && remainder100 <= 13) return `${num}th`;

  const remainder = num % 10;

  if (remainder === 1) return `${num}st`;
  if (remainder === 2) return `${num}nd`;
  if (remainder === 3) return `${num}rd`;

  return `${num}th`;
}

export function humanProcessingTime(seconds: number): string {
  return numeral(seconds).format("00:00:00");
}

export function niceNumber(value: number | string | null | undefined) {
  if (isNil(value)) return "N/A";
  if (isString(value) && isEmpty(value)) return "N/A";
  return numeral(value).format("0,0");
}

export function niceByte(value: number | string | null | undefined) {
  if (isNil(value)) return "N/A";
  if (isString(value) && isEmpty(value)) return "N/A";
  return numeral(value).format("0b");
}

class LocalPersistance {
  // https://developer.mozilla.org/en-US/docs/Web/API/Storage

  supported: boolean = typeof Storage !== "undefined";

  private wrap(fn: () => unknown): unknown {
    if (!this.supported) return;
    try {
      return fn();
    } catch (error) {
      console.error(error);
    }
  }

  clear() {
    this.wrap(() => localStorage.clear());
  }

  getItem(key: string): string | null | undefined {
    return this.wrap(() => localStorage.getItem(key)) as string | null | undefined;
  }

  setItem(key: string, value: string) {
    return this.wrap(() => localStorage.setItem(key, value));
  }

  removeItem(key: string) {
    return this.wrap(() => localStorage.removeItem(key));
  }
}

export const storage = new LocalPersistance();

export function removeSpaces(text: string): string {
  if (isEmpty(text)) return "";

  return text.replace(/^\s+|\s+$/gm,'');
}