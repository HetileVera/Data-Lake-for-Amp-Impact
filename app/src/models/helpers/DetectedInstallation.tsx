import { types, Instance } from "mobx-state-tree";

/**
 * Represents a short information about a detected installation
 */
export const DetectedInstallation = types
  .model("DetectedInstallation", {
    id: "",
    webUrlOrigin: "",
    reachable: true,
    region: "",
    installationJson: types.maybe(types.frozen()),
  })

  .views((self) => ({
    get startDate(): Date | undefined {
      return self.installationJson?.startDate;
    },

    get startedBy(): string | undefined {
      return self.installationJson?.startedBy;
    },

    get accountId(): string | undefined {
      return self.installationJson?.accountId;
    },

    get appFlowConnectionName(): string | undefined {
      return self.installationJson?.appFlowConnectionName;
    },
  }));

export interface IDetectedInstallation extends Instance<typeof DetectedInstallation> {}
