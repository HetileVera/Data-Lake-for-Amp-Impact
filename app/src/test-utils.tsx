import React from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";

import { theme } from "themes/orange";

export const CustomChakraProvider = ({ children }: { children: React.ReactNode }) => {
  return (
    <ChakraProvider theme={theme}>
      <div data-testid="test-root">{children}</div>
    </ChakraProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: Omit<RenderOptions, "wrapper">) => render(ui, { wrapper: CustomChakraProvider, ...options });

export * from "@testing-library/react";

export { customRender as render };
