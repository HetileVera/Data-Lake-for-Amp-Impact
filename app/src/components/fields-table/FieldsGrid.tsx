import React from "react";
import { SimpleGrid } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { IField } from "models/helpers/Field";

import { SelectableField } from "./SelectableField";

interface IFieldsGrid {
  fields: IField[];
}

export const FieldsGrid = observer(({ fields }: IFieldsGrid) => {
  if (fields.length < 1) return null;
  return (
    <SimpleGrid minChildWidth="200px" spacing="15px" mt={6}>
      {fields.map((field) => (
        <SelectableField key={field.name} field={field} />
      ))}
    </SimpleGrid>
  );
});
