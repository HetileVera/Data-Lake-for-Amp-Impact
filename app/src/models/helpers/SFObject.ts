import filter from "lodash/filter";
import has from "lodash/has";
import sortBy from "lodash/sortBy";

import { types, Instance } from "mobx-state-tree";
import { values } from "mobx";

import { IRawSalesforceObject } from "api/list-salesforce-objects";
import { defaultObjectsMap } from "data/default-npsp-objects";
import { Field, IField } from "models/helpers/Field";
import { isCompoundType } from "models/helpers/utils";

/**
 * Represents a Salesforce object
 */

const sfProps = {
  name: "",
  label: "",
  fields: types.map(Field),
  fieldsLoaded: false,
  selected: false,
  stale: false,
};

const sfActions = (self: any) => ({
  setObject(rawObject: IRawSalesforceObject) {
    self.fields.clear();
    self.name = rawObject.name;
    self.label = rawObject.label;
    self.fieldsLoaded = false;
    self.selected = false;
    self.stale = false;
  },

  setFields(rawFields: { name: string; label: string; type: string }[]) {
    self.fields.clear();
    for (const rawField of rawFields) {
      // Compound fields are excluded
      self.fields.set(rawField.name, { ...rawField, excluded: isCompoundType(rawField.type) });
    }
  },

  setField(rawField: { name: string; label: string; type: string; excluded: boolean }) {
    self.fields.set(rawField.name, { ...rawField, excluded: rawField.excluded || isCompoundType(rawField.type) });
  },

  setFieldsLoaded(flag: boolean) {
    self.fieldsLoaded = flag;
  },

  setSelected(flag: boolean) {
    self.selected = flag;
  },

  markStale() {
    self.stale = true;
  },
});

const sfViews = (self: any) => ({
  get selectableFields(): IField[] {
    // Returns the list of fields that can be selected
    const filtered = filter(values(self.fields), ["canExclude", true]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get listableFields(): IField[] {
    // listable fields are all the fields with the exception of the compound field
    const filtered = filter(values(self.fields), ["compound", false]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get immutableFields(): IField[] {
    const filtered = filter(values(self.fields), { compound: false, canExclude: false });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get selectedFields(): IField[] {
    // Returns all fields that are included (including immutable fields)
    const filtered = filter(values(self.fields), ["excluded", false]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get userSelectedFields(): IField[] {
    // Returns all fields that are included (excluded immutable fields)
    const filtered = filter(values(self.fields), { excluded: false, canExclude: true });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get excludedFields(): IField[] {
    // Returns excluded fields (with the exception of compound fields)
    const filtered = filter(values(self.fields), { excluded: true, compound: false });
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },

  get allExcludedFields(): IField[] {
    // Returns all excluded fields including the compound field
    const filtered = filter(values(self.fields), ["excluded", true]);
    const result: unknown[] = sortBy(filtered, ["label", "name"]);

    return result as IField[];
  },
  get fieldsCount(): number {
    // All fields (selected or not) excluding compound fields
    return self.listableFields.length;
  },

  get selectedFieldsCount(): number {
    const result = self.selectedFields;
    return result.length;
  },

  get excludedFieldsCount(): number {
    const result = self.excludedFields;
    return result.length;
  },

  get excluded() {
    // Returns the excluded fields as expected by the step functions lambdas. The following is an example of the
    // expected exclude:
    // "exclude": {
    //   "CompoundField__c": { "type": "location", "label": "<field label>", "$comment": "Compound Field" },
    //   "ABC__c": { "type": "string", "label", "<field label>", "$comment": "ABC field" },
    //   "...": { ... },
    // },
    const excludedFields = self.allExcludedFields;
    const exclude: { [index: string]: { type: string; label: string } } = {};
    for (const field of excludedFields) {
      exclude[field.name] = {
        type: field.type,
        label: field.label,
      };
    }

    return exclude;
  },

  get isDefault(): boolean {
    // Is this object one of the default NPSP objects?
    return has(defaultObjectsMap, self.name);
  },
});

export const SFObject = types.model("SFObject", sfProps).actions(sfActions).views(sfViews);

export interface ISFObject extends Instance<typeof SFObject> {}
