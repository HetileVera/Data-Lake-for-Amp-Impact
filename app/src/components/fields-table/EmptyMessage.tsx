import React from "react";
import isEmpty from "lodash/isEmpty";
import { Box } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { FieldsViewOptions } from "models/useFieldsTable";
import { ISFObject } from "models/helpers/SFObject";
import { IField } from "models/helpers/Field";

interface IEmptyMessage {
  object: ISFObject;
  viewOption: FieldsViewOptions;
  fields: IField[];
  searchText: string;
}

export const EmptyMessage = observer(({ object, viewOption, fields, searchText }: IEmptyMessage) => {
  const { tone } = useColorScheme();

  if (fields.length > 0) return null;

  let content = "No fields are found";
  let size = 0;

  if (!isEmpty(searchText)) content = "No matches are found";
  else {
    switch (viewOption) {
      case FieldsViewOptions.All:
        size = object.immutableFields.length;
        if (size > 0) {
          content = `There are ${size} included fields. They are mandatory.`;
        } else {
          content = "No fields are found";
        }
        break;
      case FieldsViewOptions.Only_Included:
        size = object.immutableFields.length;
        if (size > 0) {
          content = `There are ${size} included fields. They are mandatory.`;
        } else {
          content = "No included fields are found";
        }
        break;
      case FieldsViewOptions.Only_Excluded:
        content = "No excluded fields are found";
    }
  }

  return (
    <Box bg={tone(100)} textAlign="center" p={4} fontSize="xs" color={tone(600)} borderRadius="md">
      {content}
    </Box>
  );
});
