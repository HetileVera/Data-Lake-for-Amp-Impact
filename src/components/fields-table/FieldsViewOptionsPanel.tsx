import React from "react";
import { Text, Badge, Flex, RadioGroup, Radio } from "@chakra-ui/react";
import { observer } from "mobx-react";

import { useColorScheme } from "models/useColorScheme";
import { FieldsViewOptions } from "models/useFieldsTable";
import { ISFObject } from "models/helpers/SFObject";

interface IFieldsViewOptionsPanel {
  object: ISFObject;
  viewOption: FieldsViewOptions;
  setViewOption: (option: FieldsViewOptions) => void;
}

export const FieldsViewOptionsPanel = observer(({ object, viewOption, setViewOption }: IFieldsViewOptionsPanel) => {
  const { tone, colorScheme } = useColorScheme();
  const is = (option: FieldsViewOptions) => option === viewOption;

  return (
    <RadioGroup defaultValue="1" mb={6}>
      <Flex fontSize="sm" justifyContent="space-between" pr={2}>
        <Flex pl={1}>
          Show
          <Radio
            ml={6}
            value="1"
            colorScheme={colorScheme}
            isChecked={is(FieldsViewOptions.All)}
            onChange={() => setViewOption(FieldsViewOptions.All)}
          >
            <Text color={tone(700)} fontSize="sm" mr={4}>
              All Fields
              <Badge ml={2} bg={tone(75)} fontSize="xs" borderRadius="full" fontWeight="normal">
                {object.fieldsCount}
              </Badge>
            </Text>
          </Radio>
        </Flex>
        <Radio
          value="2"
          colorScheme={colorScheme}
          isChecked={is(FieldsViewOptions.Only_Included)}
          onChange={() => setViewOption(FieldsViewOptions.Only_Included)}
        >
          <Text color={tone(700)} fontSize="sm" mr={4}>
            Included Fields
            <Badge ml={2} bg={tone(75)} fontSize="xs" borderRadius="full" fontWeight="normal">
              {object.selectedFieldsCount}
            </Badge>
          </Text>
        </Radio>
        <Radio
          value="3"
          colorScheme={colorScheme}
          isChecked={is(FieldsViewOptions.Only_Excluded)}
          onChange={() => setViewOption(FieldsViewOptions.Only_Excluded)}
        >
          <Text fontSize="sm" color={tone(700)}>
            Excluded Fields
            <Badge ml={2} bg={tone(75)} fontSize="xs" borderRadius="full" fontWeight="normal">
              {object.excludedFieldsCount}
            </Badge>
          </Text>
        </Radio>
      </Flex>
    </RadioGroup>
  );
});
