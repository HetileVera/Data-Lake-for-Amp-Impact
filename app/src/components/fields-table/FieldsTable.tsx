import React from "react";
import isEmpty from "lodash/isEmpty";
import { Box, Flex, IconButton, InputGroup, Input, InputRightElement } from "@chakra-ui/react";
import { SmallCloseIcon } from "@chakra-ui/icons";
import { observer } from "mobx-react";

import { PaginationButtons } from "components/PaginationButtons";
import { useColorScheme } from "models/useColorScheme";
import { useFieldsTable } from "models/useFieldsTable";
import { ISFObject } from "models/helpers/SFObject";

import { FieldsViewOptionsPanel } from "./FieldsViewOptionsPanel";
import { EmptyMessage } from "./EmptyMessage";
import { FieldsGrid } from "./FieldsGrid";
import { ImmutableFieldsGrid } from "./ImmutableFieldsGrid";

interface IFieldsTable {
  object: ISFObject;
}

export const FieldsTable = observer(({ object }: IFieldsTable) => {
  const { tone, colorScheme } = useColorScheme();
  const { fields, viewOption, setViewOption, totalPages, currentPage, setCurrentPage, searchText, setSearchText, totalMatches } = useFieldsTable({
    object,
  });

  const handleTextChange: React.ChangeEventHandler<HTMLInputElement> = (event) => setSearchText(event.target.value);

  return (
    <Box p={3} borderRadius="md" borderWidth="1px" borderColor={tone(100)}>
      <FieldsViewOptionsPanel object={object} viewOption={viewOption} setViewOption={setViewOption} />

      {(totalPages > 1 || !isEmpty(searchText)) && (
        <Box position="relative" mb={8}>
          <Flex mb={0}>
            <InputGroup size="sm">
              <Input type="text" placeholder="Search" borderRadius="md" value={searchText} onChange={handleTextChange} />
              <InputRightElement>
                <IconButton
                  onClick={() => setSearchText("")}
                  size="sm"
                  variant="ghost"
                  color="purple.200"
                  colorScheme={colorScheme}
                  aria-label="Clear search text"
                  _active={{ bg: "none", color: tone(900) }}
                  _hover={{ bg: "none", color: tone(500) }}
                  _focus={{ boxShadow: "none" }}
                  icon={<SmallCloseIcon fontSize="lg" />}
                />
              </InputRightElement>
            </InputGroup>
            <PaginationButtons totalPages={totalPages} currentPage={currentPage} setCurrentPage={setCurrentPage} />
          </Flex>

          {!isEmpty(searchText) && (
            <Box p={0} m={0} fontSize="0.6rem" position="absolute" bottom="-20px" left="10px">
              Found <b>{totalMatches}</b> matches
            </Box>
          )}
        </Box>
      )}
      <EmptyMessage object={object} fields={fields} viewOption={viewOption} searchText={searchText} />
      <FieldsGrid fields={fields} />
      <ImmutableFieldsGrid object={object} viewOption={viewOption} />
    </Box>
  );
});
