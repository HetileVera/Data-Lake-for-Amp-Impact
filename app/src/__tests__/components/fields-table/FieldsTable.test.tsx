import { cleanup } from "@testing-library/react";
import { create } from "react-test-renderer";

import { CustomChakraProvider } from "test-utils";

import { FieldsTable } from "components/fields-table/FieldsTable";
import { SFObject } from "models/helpers/SFObject";

describe("FieldsTable component", () => {
  afterEach(cleanup);

  test("snapshot", () => {
    const sfObject = SFObject.create({
      name: "TestField",
      label: "Test Field",
      fields: {},
      fieldsLoaded: false,
      selected: false,
      stale: false,
    });

    const tree = create(
      <CustomChakraProvider>
        <FieldsTable object={sfObject} />
      </CustomChakraProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
