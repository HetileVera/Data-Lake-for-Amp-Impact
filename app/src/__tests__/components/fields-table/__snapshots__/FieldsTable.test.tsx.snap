// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldsTable component snapshot 1`] = `
Array [
  <div
    data-testid="test-root"
  >
    <div
      className="css-10n4rap"
    >
      <div
        className="chakra-radio-group css-1y0bycm"
        role="radiogroup"
      >
        <div
          className="css-1lgisrc"
        >
          <div
            className="css-np5eqa"
          >
            Show
          </div>
          <label
            className="chakra-radio css-1pw4d56"
            data-checked=""
          >
            <input
              checked={true}
              className="chakra-radio__input"
              id="radio-1"
              name="radio-1"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                Object {
                  "border": "0px",
                  "clip": "rect(0px, 0px, 0px, 0px)",
                  "height": "1px",
                  "margin": "-1px",
                  "overflow": "hidden",
                  "padding": "0px",
                  "position": "absolute",
                  "whiteSpace": "nowrap",
                  "width": "1px",
                }
              }
              type="radio"
              value="1"
            />
            <span
              aria-hidden={true}
              className="chakra-radio__control css-1upy29y"
              data-checked=""
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            />
            <span
              className="chakra-radio__label css-1i66d7g"
              data-checked=""
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <p
                className="chakra-text css-9f98jw"
              >
                NPSP Fields
                <span
                  className="chakra-badge css-7hgm7x"
                >
                  0
                </span>
              </p>
            </span>
          </label>
          <label
            className="chakra-radio css-fjnot9"
          >
            <input
              checked={false}
              className="chakra-radio__input"
              id="radio-1"
              name="radio-1"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                Object {
                  "border": "0px",
                  "clip": "rect(0px, 0px, 0px, 0px)",
                  "height": "1px",
                  "margin": "-1px",
                  "overflow": "hidden",
                  "padding": "0px",
                  "position": "absolute",
                  "whiteSpace": "nowrap",
                  "width": "1px",
                }
              }
              type="radio"
              value="2"
            />
            <span
              aria-hidden={true}
              className="chakra-radio__control css-1upy29y"
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            />
            <span
              className="chakra-radio__label css-1i66d7g"
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <p
                className="chakra-text css-9f98jw"
              >
                All Fields
                <span
                  className="chakra-badge css-7hgm7x"
                >
                  0
                </span>
              </p>
            </span>
          </label>
          <label
            className="chakra-radio css-1pw4d56"
          >
            <input
              checked={false}
              className="chakra-radio__input"
              id="radio-1"
              name="radio-1"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              style={
                Object {
                  "border": "0px",
                  "clip": "rect(0px, 0px, 0px, 0px)",
                  "height": "1px",
                  "margin": "-1px",
                  "overflow": "hidden",
                  "padding": "0px",
                  "position": "absolute",
                  "whiteSpace": "nowrap",
                  "width": "1px",
                }
              }
              type="radio"
              value="3"
            />
            <span
              aria-hidden={true}
              className="chakra-radio__control css-1upy29y"
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
            />
            <span
              className="chakra-radio__label css-1i66d7g"
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <p
                className="chakra-text css-bxgqra"
              >
                Excluded Fields
                <span
                  className="chakra-badge css-7hgm7x"
                >
                  0
                </span>
              </p>
            </span>
          </label>
        </div>
      </div>
      <div
        className="css-1y07nh"
      >
        No fields are found
      </div>
    </div>
  </div>,
  <span />,
]
`;
